File: surf-cloud-api/src/main/java/dev/slne/surf/cloud/api/config/properties/SystemProperty.java
package dev.slne.surf.cloud.api.config.properties;

import java.util.function.Function;
import org.jetbrains.annotations.ApiStatus.NonExtendable;
import org.jetbrains.annotations.Nullable;

@NonExtendable
public interface SystemProperty<T> {

  T value();

  static <T> SystemProperty<T> property(
      String key,
      Function<String, T> parser,
      @Nullable T defaultValue
  ) {
    return property(String.join(".", "surf", "cloud"), key, parser, defaultValue);
  }

  static <T> SystemProperty<T> property(
      String prefix,
      String key,
      Function<String, T> parser,
      @Nullable T defaultValue
  ) {
    return SystemPropertyImpl.property(prefix + "." + key, parser, defaultValue);
  }
}

File: surf-cloud-api/src/main/kotlin/dev/slne/surf/cloud/api/config/properties/SystemProperty.kt
package dev.slne.surf.cloud.api.config.properties

import org.jetbrains.annotations.ApiStatus

@ApiStatus.NonExtendable
interface SystemProperty<T> {
    fun value(): T?
}

fun <T> systemProperty(
    key: String,
    parser: (String) -> T,
    defaultValue: T?
): SystemProperty<T> = SystemPropertyImpl(
    arrayOf("surf", "cloud", key).joinToString("."),
    parser,
    defaultValue
)


fun <T> systemProperty(
    prefix: String,
    key: String,
    parser: (String) -> T,
    defaultValue: T?
): SystemProperty<T> = SystemPropertyImpl("$prefix.$key", parser, defaultValue)

